requests/models.py

```diff
@@ -455,28 +443,39 @@ def prepare_body(self, data, files, json=None):
                     else:
                         content_type = 'application/x-www-form-urlencoded'
 
-            self.prepare_content_length(body)
-
             # Add content-type if it wasn't explicitly provided.
             if content_type and ('content-type' not in self.headers):
                 self.headers['Content-Type'] = content_type
 
+        self.prepare_content_length(body)
         self.body = body
 
     def prepare_content_length(self, body):
-        if hasattr(body, 'seek') and hasattr(body, 'tell'):
-            curr_pos = body.tell()
-            body.seek(0, 2)
-            end_pos = body.tell()
-            self.headers['Content-Length'] = builtin_str(max(0, end_pos - curr_pos))
-            body.seek(curr_pos, 0)
-        elif body is not None:
-            l = super_len(body)
-            if l:
-                self.headers['Content-Length'] = builtin_str(l)
+        """Prepares Content-Length header.
+        
+        If the length of the body of the request can be computed, Content-Length is set using
+        super_len. If user has manually set either a Transfer-Encoding or Content-Length header
+        when it should not be set (they should be mutually exclusive) an InvalidHeaderError
+        error will be raised.
+        """
+        if body is not None:
+            is_stream = determine_if_stream(body)
+
+            try:
+                length = super_len(body)
+            except (TypeError, AttributeError, UnsupportedOperation):
+                length = None
+
+            if length:
+                self.headers['Content-Length'] = builtin_str(length)
+            elif is_stream and not length:
+                self.headers['Transfer-Encoding'] = 'chunked'
```

Comment:
Our point is that what happens when it's something without a length, that is not streaming. That seems to be a corner case not presently handled (at least not explicitly). We kind of handle it in the next elif but only for certain methods and I'm starting to think we just need a comment explaining why that handling is okay.